flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java

```diff
@@ -412,92 +492,267 @@ public static Long toTimestamp(String dateStr, TimeZone tz) {
+                + milli;
     }
 
     /**
-     * Format a timestamp as specific.
+     * Parses a fraction, multiplying the first character by {@code multiplier}, the second
+     * character by {@code multiplier / 10}, the third character by {@code multiplier / 100}, and so
+     * forth.
      *
-     * @param ts the timestamp to format.
-     * @param format the string formatter.
-     * @param tz the time zone
+     * <p>For example, {@code parseFraction("1234", 100)} yields {@code 123}.
      */
-    public static String dateFormat(long ts, String format, TimeZone tz) {
-        SimpleDateFormat formatter = FORMATTER_CACHE.get(format);
-        formatter.setTimeZone(tz);
-        Date dateTime = new Date(ts);
-        return formatter.format(dateTime);
-    }
-
+    private static int parseFraction(String v) {
+        int multiplier = 100;
+        int r = 0;
+        for (int i = 0; i < v.length(); i++) {
+            char c = v.charAt(i);
+            int x = c < '0' || c > '9' ? 0 : (c - '0');
+            r += multiplier * x;
+            if (multiplier < 10) {
+                // We're at the last digit. Check for rounding.
+                if (i + 1 < v.length() && v.charAt(i + 1) >= '5') {
+                    ++r;
+                }
+                break;
+            }
+            multiplier /= 10;
+        }
+        return r;
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Format
+    // --------------------------------------------------------------------------------------------
+
+    public static String formatTimestamp(TimestampData ts, String format) {
+        return formatTimestamp(ts, format, ZoneId.of("UTC"));
+    }
+
+    public static String formatTimestamp(TimestampData ts, String format, TimeZone zone) {
+        return formatTimestamp(ts, format, zone.toZoneId());
+    }
+
+    private static String formatTimestamp(TimestampData ts, int precision) {
+        LocalDateTime ldt = ts.toLocalDateTime();
+
+        String fraction = pad(9, ldt.getNano());
+        while (fraction.length() > precision && fraction.endsWith("0")) {
+            fraction = fraction.substring(0, fraction.length() - 1);
+        }
+
+        StringBuilder ymdhms =
+                ymdhms(
+                        new StringBuilder(),
+                        ldt.getYear(),
+                        ldt.getMonthValue(),
+                        ldt.getDayOfMonth(),
+                        ldt.getHour(),
+                        ldt.getMinute(),
+                        ldt.getSecond());
+
+        if (fraction.length() > 0) {
+            ymdhms.append(".").append(fraction);
+        }
+
+        return ymdhms.toString();
+    }
+
+    public static String formatTimestamp(TimestampData ts, TimeZone tz, int precision) {
+        return formatTimestamp(timestampWithLocalZoneToTimestamp(ts, tz), precision);
+    }
+
+    /**
+     * Format a timestamp as specific.
```

Comment:
same here, I think the javadoc is superfluous.

Reply:
Removed javadoc here as well